File: androidApp/features/taskdetail/src/main/kotlin/com/velkonost/getbetter/android/features/taskdetail/TaskDetailScreen.kt
package com.velkonost.getbetter.android.features.taskdetail

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.ModalBottomSheetValue
import androidx.compose.material.Text
import androidx.compose.material.rememberModalBottomSheetState
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.velkonost.getbetter.android.features.areadetail.AreaDetailScreen
import com.velkonost.getbetter.android.features.taskdetail.components.AbilityData
import com.velkonost.getbetter.android.features.taskdetail.components.AbilityDataHidden
import com.velkonost.getbetter.android.features.taskdetail.components.TaskDetailHeader
import com.velkonost.getbetter.core.compose.components.Loader
import com.velkonost.getbetter.core.compose.components.createnewnote.CreateNewNoteBottomSheet
import com.velkonost.getbetter.core.compose.components.details.AreaData
import com.velkonost.getbetter.core.compose.components.notelist.AddNoteItem
import com.velkonost.getbetter.core.compose.components.notelist.NoteItem
import com.velkonost.getbetter.core.compose.extensions.OnBottomReached
import com.velkonost.getbetter.shared.features.createnote.presentation.contract.CreateNewNoteAction
import com.velkonost.getbetter.shared.features.taskdetail.presentation.TaskDetailViewModel
import com.velkonost.getbetter.shared.features.taskdetail.presentation.contract.TaskDetailAction
import com.velkonost.getbetter.shared.features.taskdetail.presentation.contract.TaskDetailEvent
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.stringResource
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun TaskDetailScreen(
    modifier: Modifier = Modifier,
    viewModel: TaskDetailViewModel
) {

    val state by viewModel.viewState.collectAsStateWithLifecycle()
    val scope = rememberCoroutineScope()

    val interactionSource = remember { MutableInteractionSource() }
    val listState = rememberLazyListState()

    val selectedAreaId = remember { mutableStateOf<Int?>(null) }
    val areaDetailSheetState = rememberModalBottomSheetState(
        initialValue = ModalBottomSheetValue.Hidden,
        skipHalfExpanded = true
    )

    val createNewNoteSheetState = rememberModalBottomSheetState(
        initialValue = ModalBottomSheetValue.Hidden,
        skipHalfExpanded = true,
    )

    Box(modifier = modifier.fillMaxSize()) {
        if (state.isLoading || state.task == null) {
            Loader(modifier = Modifier.align(Alignment.Center))
        } else {
            LazyColumn(
                modifier = modifier
                    .padding(start = 20.dp, end = 20.dp)
                    .fillMaxSize(),
                state = listState,
                contentPadding = PaddingValues(bottom = 220.dp)
            ) {
                item {
                    TaskDetailHeader(
                        isShortInfo = state.task!!.isShortInfo,
                        isFavorite = state.task!!.isFavorite,
                        isFavoriteLoading = state.task!!.isFavoriteLoading,
                        onBackClick = {
                            viewModel.dispatch(TaskDetailAction.NavigateBack)
                        },
                        onFavoriteClick = {
                            viewModel.dispatch(TaskDetailAction.FavoriteClick)
                        }
                    )
                }

                item {
                    state.area?.let {
                        AreaData(area = it) {
                            scope.launch {
                                selectedAreaId.value = it.id
                                areaDetailSheetState.show()
                            }
                        }
                    }
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 24.dp),
                        text = state.task!!.name,
                        style = MaterialTheme.typography.labelLarge.copy(fontSize = 18.sp),
                        color = colorResource(resource = SharedR.colors.text_light),
                    )
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 16.dp),
                        text = stringResource(resource = SharedR.strings.task_what_to_do_title),
                        style = MaterialTheme.typography.labelMedium,
                        color = colorResource(resource = SharedR.colors.text_title)
                    )
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 6.dp),
                        text = state.task!!.whatToDo,
                        style = MaterialTheme.typography.labelMedium,
                        color = colorResource(resource = SharedR.colors.text_primary)
                    )
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 16.dp),
                        text = stringResource(resource = SharedR.strings.task_why_title),
                        style = MaterialTheme.typography.labelMedium,
                        color = colorResource(resource = SharedR.colors.text_title)
                    )
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 6.dp),
                        text = state.task!!.why,
                        style = MaterialTheme.typography.labelMedium,
                        color = colorResource(resource = SharedR.colors.text_primary)
                    )
                }

                item {
                    Text(
                        modifier = modifier.padding(top = 16.dp),
                        text = stringResource(resource = SharedR.strings.task_abilities_title),
                        style = MaterialTheme.typography.labelMedium,
                        color = colorResource(resource = SharedR.colors.text_title)
                    )
                }

                if (!state.task!!.isShortInfo) {
                    items(state.task!!.abilities, key = { it.id!! }) { ability ->
                        AbilityData(
                            item = ability,
                            onClick = {

                            }
                        )
                    }

                    item {
                        Row {
                            Spacer(modifier.weight(1f))
                            Text(
                                modifier = modifier.padding(top = 24.dp),
                                text = stringResource(resource = SharedR.strings.task_mask_as).uppercase(),
                                style = MaterialTheme.typography.labelMedium,
                                color = colorResource(resource = SharedR.colors.text_primary)
                            )
                            Spacer(modifier.weight(1f))
                        }
                    }

                    item {
                        Row(modifier = modifier.padding(top = 12.dp)) {
                            Spacer(modifier.weight(1f))
                            Spacer(modifier.weight(1f))

                            Text(
                                modifier = modifier
                                    .padding(end = 6.dp)
                                    .shadow(
                                        elevation = (if (state.task!!.isNotInteresting) 8 else 0).dp,
                                        shape = MaterialTheme.shapes.medium,
                                    )
                                    .background(
                                        color = colorResource(
                                            resource = if (state.task!!.isNotInteresting) SharedR.colors.button_gradient_start
                                            else SharedR.colors.background_item
                                        ),
                                        shape = MaterialTheme.shapes.medium
                                    )
                                    .padding(horizontal = 16.dp, vertical = 6.dp)
                                    .clickable(
                                        interactionSource = interactionSource,
                                        indication = null
                                    ) { viewModel.dispatch(TaskDetailAction.NotInterestingClick) },
                                text = stringResource(resource = SharedR.strings.task_not_interesting_title),
                                style = MaterialTheme.typography.labelLarge,
                                color = colorResource(
                                    resource = if (state.task!!.isNotInteresting) SharedR.colors.text_light
                                    else SharedR.colors.text_primary
                                )
                            )
                            Text(
                                modifier = modifier
                                    .padding(start = 6.dp)
                                    .shadow(
                                        elevation = (if (state.task!!.isCompleted) 8 else 0).dp,
                                        shape = MaterialTheme.shapes.medium,
                                    )
                                    .background(
                                        color = colorResource(
                                            resource = if (state.task!!.isCompleted) SharedR.colors.button_gradient_start
                                            else SharedR.colors.background_item
                                        ),
                                        shape = MaterialTheme.shapes.medium
                                    )
                                    .padding(horizontal = 16.dp, vertical = 6.dp)
                                    .clickable(
                                        interactionSource = interactionSource,
                                        indication = null
                                    ) { viewModel.dispatch(TaskDetailAction.CompletedClick) },
                                text = stringResource(resource = SharedR.strings.task_completed_title),
                                style = MaterialTheme.typography.labelLarge,
                                color = colorResource(
                                    resource = if (state.task!!.isCompleted) SharedR.colors.text_light
                                    else SharedR.colors.text_primary
                                )
                            )

                            Spacer(modifier.weight(1f))
                            Spacer(modifier.weight(1f))
                        }
                    }
                } else {
                    item {
                        AbilityDataHidden()
                    }
                }

                item {
                    AnimatedVisibility(visible = state.userNotesViewState.items.isNotEmpty()) {
                        androidx.compose.material3.Text(
                            modifier = modifier
                                .padding(top = 24.dp)
                                .fillMaxWidth(0.8f),
                            text = stringResource(resource = SharedR.strings.task_user_notes_title),
                            color = colorResource(resource = SharedR.colors.text_primary),
                            style = MaterialTheme.typography.headlineSmall
                        )
                    }
                }

                if (state.userNotesViewState.items.isEmpty() && state.userNotesViewState.isLoading) {
                    item {
                        Loader(modifier = Modifier
                            .fillMaxWidth()
                            .align(Alignment.Center))
                    }
                }

                items(state.userNotesViewState.items, key = { it.id }) {
                    NoteItem(
                        item = it,
                        horizontalPadding = 0,
                        onClick = {

                        },
                        onLikeClick = {

                        }
                    )
                }
            }

            if (!state.task!!.isShortInfo) {
                AddNoteItem(
                    paddingBottom = 70,
                    createGoalClick = {
                        scope.launch {
                            if (state.createNewNoteViewState.availableAreas.isEmpty()) {
                                viewModel.dispatch(CreateNewNoteAction.CloseBecauseZeroAreas)
                            } else {
                                viewModel.dispatch(TaskDetailAction.CreateGoalClick)
                                createNewNoteSheetState.show()
                            }
                        }
                    },
                    createNoteClick = {
                        scope.launch {
                            if (state.createNewNoteViewState.availableAreas.isEmpty()) {
                                viewModel.dispatch(CreateNewNoteAction.CloseBecauseZeroAreas)
                            } else {
                                viewModel.dispatch(TaskDetailAction.CreateNoteClick)
                                createNewNoteSheetState.show()
                            }
                        }
                    }
                )
            }
        }
    }

    listState.OnBottomReached(
        buffer = state.userNotesViewState.loadMorePrefetch,
        isLoading = state.userNotesViewState.isLoading
    ) {
        viewModel.dispatch(TaskDetailAction.UserNotesLoadNextPage)
    }

    AreaDetailScreen(
        modalSheetState = areaDetailSheetState,
        areaId = selectedAreaId.value,
        onAreaChanged = {
            viewModel.dispatch(TaskDetailAction.AreaChanged)
        }
    )

    CreateNewNoteBottomSheet(
        state = state.createNewNoteViewState,
        modalSheetState = createNewNoteSheetState,
        onAreaSelect = {
            viewModel.dispatch(CreateNewNoteAction.AreaSelect(it))
        },
        onTaskSelect = {
            viewModel.dispatch(CreateNewNoteAction.TaskSelect(it))
        },
        onTextChanged = {
            viewModel.dispatch(CreateNewNoteAction.TextChanged(it))
        },
        onPrivateChanged = {
            viewModel.dispatch(CreateNewNoteAction.PrivateChanged)
        },
        onNewTagChanged = {
            viewModel.dispatch(CreateNewNoteAction.NewTagTextChanged(it))
        },
        onAddNewTag = {
            viewModel.dispatch(CreateNewNoteAction.AddNewTag)
        },
        onTagDelete = {
            viewModel.dispatch(CreateNewNoteAction.RemoveTag(it))
        },
        onNewSubNoteChanged = {
            viewModel.dispatch(CreateNewNoteAction.NewSubNoteTextChanged(it))
        },
        onAddNewSubNote = {
            viewModel.dispatch(CreateNewNoteAction.AddSubNote)
        },
        onSubNoteDelete = {
            viewModel.dispatch(CreateNewNoteAction.RemoveSubNote(it))
        },
        onSetCompletionDate = {
            viewModel.dispatch(CreateNewNoteAction.SetCompletionDate(it))
        },
        onCreateClick = {
            viewModel.dispatch(CreateNewNoteAction.CreateClick)
        }
    )

    LaunchedEffect(Unit) {
        viewModel.events.collectLatest {
            when (it) {
                is TaskDetailEvent.NewNoteCreatedSuccess -> {
                    createNewNoteSheetState.hide()
                }
            }
        }
    }

}

File: iosApp/iosApp/Features/TaskDetail/TaskDetailScreen.swift
//
//  TaskDetailScreen.swift
//  iosApp
//
//  Created by velkonost on 03.12.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK
import KMMViewModelSwiftUI
import KMPNativeCoroutinesAsync

struct TaskDetailScreen : View {
    
    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
    
    @StateViewModel var viewModel: TaskDetailViewModel
    @State private var eventsObserver: Task<(), Error>? = nil
    
    @State private var selectedAreaId: Int32? = nil
    @State private var showingAreaDetailSheet = false
    @State private var showingCreateNewNoteSheet = false
    
    var body: some View {
        @State var state = viewModel.viewStateValue as! TaskDetailViewState
        @State var createNewNoteState = state.createNewNoteViewState
        
        ZStack {
            if state.isLoading || state.task == nil {
                VStack {
                    Loader()
                }
            } else {
                ScrollView(.vertical, showsIndicators: false) {
                    ScrollViewReader { value in
                        LazyVStack(spacing: 0) {
                            TaskDetailHeader(
                                isShortInfo: state.task!.isShortInfo,
                                isFavorite: state.task!.isFavorite,
                                isFavoriteLoading: state.task!.isFavoriteLoading,
                                onFavoriteClick: {
                                    viewModel.dispatch(action: TaskDetailActionFavoriteClick())
                                }
                            )
                            
                            if state.area != nil {
                                AreaData(
                                    area: state.area!,
                                    onClick: {
                                        selectedAreaId = state.area!.id
                                        showingAreaDetailSheet = true
                                    }
                                )
                            }
                            
                            HStack {
                                Text(state.task!.name)
                                    .style(.labelLarge, withSize: 18)
                                    .foregroundColor(.textLight)
                                    .padding(.top, 24)
                                Spacer()
                            }
                            
                            HStack {
                                Text(SharedR.strings().task_what_to_do_title.desc().localized())
                                    .style(.labelMedium)
                                    .foregroundColor(.textTitle)
                                    .padding(.top, 16)
                                Spacer()
                            }
                            
                            HStack {
                                Text(state.task!.whatToDo)
                                    .style(.labelMedium)
                                    .foregroundColor(.textPrimary)
                                    .padding(.top, 6)
                                Spacer()
                            }
                            
                            HStack {
                                Text(SharedR.strings().task_why_title.desc().localized())
                                    .style(.labelMedium)
                                    .foregroundColor(.textTitle)
                                    .padding(.top, 16)
                                Spacer()
                            }
                            
                            HStack {
                                Text(state.task!.why)
                                    .style(.labelMedium)
                                    .foregroundColor(.textPrimary)
                                    .padding(.top, 6)
                                Spacer()
                            }
                            
                            HStack {
                                Text(SharedR.strings().task_abilities_title.desc().localized())
                                    .style(.labelMedium)
                                    .foregroundColor(.textTitle)
                                    .padding(.top, 16)
                                Spacer()
                            }
                            
                            if !state.task!.isShortInfo {
                                ForEach(state.task!.abilities, id: \.self.id ) { ability in
                                    AbilityData(
                                        item: ability,
                                        onClick: { value in
                                            
                                        }
                                    )
                                }
                                
                                Text(SharedR.strings().task_mask_as.desc().localized().uppercased())
                                    .style(.labelMedium)
                                    .foregroundColor(.textPrimary)
                                    .padding(.top, 24)
                                
                                HStack(spacing: 0) {
                                    Text(SharedR.strings().task_not_interesting_title.desc().localized())
                                        .style(.labelLarge)
                                        .foregroundColor(state.task!.isNotInteresting ? .textLight : .textPrimary)
                                        .padding(.horizontal, 16)
                                        .padding(.vertical, 8)
                                        .background(
                                            RoundedRectangle(cornerRadius: 12)
                                                .fill(state.task!.isNotInteresting ? Color.buttonGradientStart : Color.backgroundItem)
                                                .shadow(radius: state.task!.isNotInteresting ? 8 : 0)
                                        )
                                        .padding(.trailing, 6)
                                        .onTapGesture {
                                            viewModel.dispatch(action: TaskDetailActionNotInterestingClick())
                                        }
                                    
                                    Text(SharedR.strings().task_completed_title.desc().localized())
                                        .style(.labelLarge)
                                        .foregroundColor(state.task!.isCompleted ? .textLight : .textPrimary)
                                        .padding(.horizontal, 16)
                                        .padding(.vertical, 8)
                                        .background(
                                            RoundedRectangle(cornerRadius: 12)
                                                .fill(state.task!.isCompleted ? Color.buttonGradientStart : Color.backgroundItem)
                                                .shadow(radius: state.task!.isCompleted ? 8 : 0)
                                        )
                                        .padding(.leading, 6)
                                        .onTapGesture {
                                            viewModel.dispatch(action: TaskDetailActionCompletedClick())
                                        }
                                }.padding(.top, 12)
                            } else {
                                AbilityDataHidden()
                            }
                            
                            if !state.userNotesViewState.items.isEmpty {
                                HStack {
                                    Text(SharedR.strings().task_user_notes_title.desc().localized())
                                        .style(.headlineSmall)
                                        .foregroundColor(.textPrimary)
                                    Spacer()
                                }
                                .padding(.top, 24)
                            }
                            
                            ForEach(state.userNotesViewState.items, id: \.self.id) { item in
                                NoteItem(
                                    item: item,
                                    onClick: { value in
                                        
                                    },
                                    onLikeClick: { value in
                                        
                                    }
                                )
                                .onAppear {
                                    checkPaginationThreshold(
                                        currentItemId: item.id,
                                        loadMorePrefetch: Int(state.userNotesViewState.loadMorePrefetch),
                                        isLoading: state.userNotesViewState.isLoading,
                                        onBottomReach: {
                                            viewModel.dispatch(action: TaskDetailActionUserNotesLoadNextPage())
                                        }
                                    )
                                }
                            }
                            
                            if state.userNotesViewState.items.isEmpty && state.userNotesViewState.isLoading {
                                Loader().frame(alignment: .center)
                            }
                            
                            
                            Spacer().frame(height: 140)
                        }
                        .frame(alignment: .leading)
                        .padding(.horizontal, 20)
                    }
                }
                .animation(.easeInOut, value: state.userNotesViewState)
                .onAppear {
                    if state.userNotesViewState.items.isEmpty {
                        viewModel.dispatch(action: TaskDetailActionUserNotesLoadNextPage())
                    }
                }
                
                if !state.task!.isShortInfo {
                    VStack(alignment: .trailing) {
                        Spacer()
                        AddNoteItem(
                            paddingBottom: 45,
                            createGoalClick: {
                                if state.createNewNoteViewState.availableAreas.isEmpty {
                                    viewModel.dispatch(action_: CreateNewNoteActionCloseBecauseZeroAreas())
                                } else {
                                    viewModel.dispatch(action: TaskDetailActionCreateGoalClick())
                                    showingCreateNewNoteSheet = true
                                }
                            },
                            createNoteClick: {
                                if state.createNewNoteViewState.availableAreas.isEmpty {
                                    viewModel.dispatch(action_: CreateNewNoteActionCloseBecauseZeroAreas())
                                } else {
                                    viewModel.dispatch(action: TaskDetailActionCreateNoteClick())
                                    showingCreateNewNoteSheet = true
                                }
                            }
                        )
                    }
                }
            }
        }
        .sheet(isPresented: $showingAreaDetailSheet) {
            AreaDetailScreen(
                areaId: $selectedAreaId,
                onClose: {
                    showingAreaDetailSheet = false
                },
                onAreaChanged: { areaId in
                    viewModel.dispatch(action: TaskDetailActionAreaChanged())
                }
            )
        }
        .sheet(isPresented: $showingCreateNewNoteSheet) {
            CreateNewNoteBottomSheet(
                state: $createNewNoteState,
                onAreaSelect: { area in
                    viewModel.dispatch(action_: CreateNewNoteActionAreaSelect(value: area))
                },
                onTaskSelect: { task in
                    viewModel.dispatch(action_: CreateNewNoteActionTaskSelect(value: task))
                },
                onTextChanged: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionTextChanged(value: value))
                },
                onPrivateChanged: {
                    viewModel.dispatch(action_: CreateNewNoteActionPrivateChanged())
                },
                onNewTagChanged: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionNewTagTextChanged(value: value))
                },
                onAddNewTag: {
                    viewModel.dispatch(action_: CreateNewNoteActionAddNewTag())
                },
                onTagDelete: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionRemoveTag(value: value))
                },
                onNewSubNoteChanged: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionNewSubNoteTextChanged(value: value))
                },
                onAddNewSubNote: {
                    viewModel.dispatch(action_: CreateNewNoteActionAddSubNote())
                },
                onSubNoteDelete: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionRemoveSubNote(value: value))
                },
                onSetCompletionDate: { value in
                    viewModel.dispatch(action_: CreateNewNoteActionSetCompletionDate(value: value != nil ? KotlinLong(value: value!) : nil))
                },
                onCreateClick: {
                    viewModel.dispatch(action_: CreateNewNoteActionCreateClick())
                }
            )
        }
        .onAppear {
            observeEvents()
        }
        .onDisappear {
            eventsObserver?.cancel()
            eventsObserver = nil
        }
        
    }
}

extension TaskDetailScreen {
    func observeEvents() {
        if eventsObserver == nil {
            eventsObserver = Task {
                for try await event in asyncSequence(for: viewModel.events) {
                    switch(event) {
                    case _ as TaskDetailEventNewNoteCreatedSuccess: do {
                        showingCreateNewNoteSheet = false
                    }
                    default:
                        break
                    }
                }
            }
        }
    }
    
    func checkPaginationThreshold(currentItemId: Int32, loadMorePrefetch: Int, isLoading: Bool, onBottomReach: () -> Void) {
        let data = (viewModel.viewStateValue as! TaskDetailViewState).userNotesViewState.items
        let thresholdIndex = data.index(data.endIndex, offsetBy: -loadMorePrefetch)
        
        if data.firstIndex(where: { $0.id == currentItemId })! >= thresholdIndex && !isLoading {
            onBottomReach()
        }
    }
}
